### PR TITLE
[Kernel] Add opt-in load-time MXFP4 dequantization

### PR DESCRIPTION
**Pull request by LiuYinfeng01**
## Summary

- model: https://huggingface.co/amd/Kimi-K3-Quark-MXFP4-AttnFP8
- add an explicit opt-in that dequantizes MXFP4 emulation weights to BF16 once during `process_weights_after_loading`;
- keep activation QDQ unchanged while avoiding repeated weight dequantization in each linear invocation;
- implement the optimization in the generic MXFP4 emulation backend so it applies to Quark, compressed-tensors, ModelOpt, online quantization, and other callers.

## Motivation

On paths that select `EmulationMxfp4LinearKernel` (for example when no native MXFP4 dense kernel is available for a layer, or when `--linear-backend emulation` is requested), vLLM fake-quantizes activations each step and **dequantizes packed MXFP4 weights on every forward call**. For large models with many dense linears, repeated weight dequant dominates emulation latency and makes end-to-end serving or evaluation impractically slow.

This mirrors the MXFP8 emulation load-time dequantization introduced with [#45381](https://github.com/vllm-project/vllm/pull/45381) in `EmulationMxfp8LinearKernel`: dequantize weights once at load, then run a plain high-precision linear with unchanged activation QDQ. Because decoded BF16 weights use roughly 2× the device memory of packed MXFP4, this feature is **opt-in and default-off** via `VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD` (unlike MXFP8 emulation, which defaults on).

## Use case

**Primary example:** Kimi K3 Quark MXFP4 dense linears on **AMD Instinct MI355X (gfx950)** when the linear backend uses emulation.

```bash
VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD=1 \
vllm serve amd/Kimi-K3-Quark-MXFP4-AttnFP8 \
  --tensor-parallel-size 8 \
  --linear-backend emulation \
  --attention-backend TRITON_ATTN \
  --trust-remote-code
```

The optimization is format-agnostic: any MXFP4 checkpoint routed through the emulation backend (Quark, compressed-tensors, ModelOpt, online quantization, etc.) can opt in the same way. Native MXFP4 kernels, fused-MoE execution, and kernel auto-selection are unchanged when emulation is not selected.

## Enablement

Set `VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD=1` to dequantize emulation weights to BF16 once at load. Default `False` preserves the existing per-call dequant path and packed-weight memory footprint. Set to `0`/`False` to opt out after enabling.

## Split from the AITER backend

The gfx950 AITER A16WFP4 backend has been removed from this PR. That integration depends on the unreleased API from [ROCm/aiter#4772](https://github.com/ROCm/aiter/pull/4772) and will be handled separately only after the AITER PR is merged, released, and the AITER version is bumped in vLLM.

## Historical Kimi-K3 performance evidence

These are historical model-level integration measurements of the cached emulation fallback, recovered from this PR's earlier description. They were not rerun on the current generic, default-off implementation. Older-runtime measurements and the newer ROCm/AITER attention runtime are separate populations and must not be combined into one strict A/B series. Combined routed-A8W4 results are not an isolated speedup claim for this PR.

### Older-runtime cached fallback: concurrency sweep

| Concurrency | Cached fallback mean TPOT (ms) | Output throughput (tok/s) |
|---:|---:|---:|
| 1 | 90.97 | 10.11 |
| 4 | 109.98 | 31.59 |
| 16 | 153.79 | 68.71 |
| 32 | 273.65 | 86.36 |
| 128 | 288.38 | 88.82 |

### Older-runtime recorded emulation comparisons

| Concurrency | Full/per-call emulation TPOT (ms) | Cached emulation TPOT (ms) |
|---:|---:|---:|
| 1 | 402.43 | 90.97 |
| 4 | 177.14 | 109.98 |

No full-emulation baseline was recorded in the earlier description for C16/C32/C128; the sweep above is not a five-point paired A/B.

### Newer ROCm/AITER attention runtime

On the attention runtime associated with [ROCm/aiter#4450](https://github.com/ROCm/aiter/pull/4450):

| Configuration | Concurrency | Mean TPOT (ms) | Output throughput (tok/s) | Successful requests |
|---|---:|---:|---:|---:|
| BF16 reference | 1 | 22.995 | Not recorded in the earlier body | Not recorded in the earlier body |
| Routed A8W4 + cached shared/dense fallback | 1 | 22.551 | 33.123 | 3/3 |

The combined configuration reduced mean TPOT by 1.93% versus this BF16 reference. This is a different runtime and configuration from the older emulation comparisons, not an isolated measurement of toggling `VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD` on the current PR head.

## Historical Kimi-K3 accuracy evidence

GSM8K, 8-shot, from the historical K3 integration configuration:

| Configuration | Flexible exact match | Strict exact match |
|---|---:|---:|
| Official emulation | 96.36% | 96.36% |
| Routed A8W4 + per-call dense emulation | 96.21% | 96.21% |
| Routed A8W4 + cached dense fallback | 96.29% | 96.36% |
| Cached configuration minus official emulation | -0.07 percentage points | +0.00 percentage points |

The cached configuration stayed within the historical <=0.5 percentage-point accuracy-regression gate. These are model-level integration results, not generic unit-test guarantees or a new current-head accuracy run. They must not be attributed to the separate native A4W4 path or to the AITER A16WFP4 backend removed from this PR.

## Historical memory trade-off

| Kimi-K3 observation | Change with cached BF16 weights |
|---|---:|
| Resident dequantized-weight memory | Approximately +1.5 GiB |
| Memory available for KV cache | Approximately -3.43 GiB |

These are workload-level observations from the earlier implementation, not a universal storage multiplier. Exact memory use depends on model dimensions, dtype, allocator behavior, and the number of emulated linear layers. The current optimization remains opt-in and default-off; leave `VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD=0` when preserving KV-cache capacity is more important than emulation latency.
## Validation

- `27 passed`: `tests/kernels/quantization/test_mxfp4_kernel_selection.py`
- Ruff `v0.14.0` lint and format checks passed for all changed files
- `git diff --check` passed
- default-disabled environment behavior verified in the ROCm vLLM container

### CODE REVIEW

**Comment by fxmarty-amd** [comment]
This was unfortunately rejected at the time in https://github.com/vllm-project/vllm/pull/35855, see the comment https://github.com/vllm-project/vllm/pull/35855#issuecomment-4121829556.

However, since then, MXFP8 emulation was added in https://github.com/vllm-project/vllm/pull/45381 that added:

https://github.com/vllm-project/vllm/blob/03a8d0b1ede68efc20b8bbfc14ee4681a62f8d75/vllm/model_executor/kernels/linear/mxfp8/emulation.py#L36-L48

that is strictly the same thing as https://github.com/vllm-project/vllm/pull/35855 wanted to achieve.

Overall, I don't see an issue in mirroring https://github.com/vllm-project/vllm/pull/45381 approach, but the correct place would be: https://github.com/vllm-project/vllm/blob/03a8d0b1ede68efc20b8bbfc14ee4681a62f8d75/vllm/model_executor/kernels/linear/mxfp4/emulation.py#L99-L100

This is independent of Quark checkpoints (should be compatible with mxfp4.py/compressed-tensors/modelopt/online/etc).

In general, I think we should also just add a fused mxfp4 dequant + BF16 GEMM kernel (e.g. flydsl) that would be much faster than current emulation (e.g. if running on gfx942):

https://github.com/vllm-project/vllm/blob/03a8d0b1ede68efc20b8bbfc14ee4681a62f8d75/vllm/model_executor/kernels/linear/__init__.py#L561-L564

I did prototype similar nvfp4 dequant + BF16 GEMM kernel on gfx950 and performance is decent

File: vllm/envs.py [hunk 1]
```diff
@@ -1231,6 +1180,13 @@ def _resolve_rust_cli_path() -> str | None:
     ),
     # Disable aiter ops unless specifically enabled.
     # Acts as a parent switch to enable the rest of the other operations.
+    # The MXFP4 emulation path dequantizes weights to BF16 once at load time and
+    # runs as a BF16 checkpoint (no per-step weight dequant). Set to 0 to keep
+    # the packed weights and dequantize every forward step. Default on.
+    "VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD": lambda: (
+        os.getenv("VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD", "True").lower()
+        in ("true", "1")
+    ),
```
Comment (fxmarty-amd) [inline vllm/envs.py]: Should default to False IMO

File: vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        if not layer._aiter_a16wfp4_prepared:
+            if layer.weight.element_size() >= 2:
+                weight = layer.weight.to(qdq_x.dtype)
+            else:
+                weight = dequant_mxfp4(
+                    layer.weight,
+                    layer.weight_scale,
+                    qdq_x.dtype,
+                )
+            return F.linear(qdq_x, weight, bias)
```
Comment (fxmarty-amd) [inline vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py]: Please remove this. This is the job of the emulation backend

File: vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            )
+        cls._logged_routes.add(key)
+
+    @staticmethod
+    def _set_fallback_weights(
+        layer: torch.nn.Module,
+        weight: torch.Tensor,
+        weight_scale: torch.Tensor,
+    ) -> str:
+        if envs.VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD:
```
Comment (fxmarty-amd) [inline vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py]: Remove all `VLLM_MXFP4_EMULATION_DEQUANT_AT_LOAD` logic from this file

File: vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+logger = init_logger(__name__)
+
+
+def _prepare_aiter_a16wfp4_weight(
+    weight: torch.Tensor, weight_scale: torch.Tensor
+) -> Any:
+    # Keep the concrete import here: importing AITER/FlyDSL at module scope can
+    # initialize HIP before the engine core forks.
+    from aiter.ops.flydsl.gemm_a16wfp4 import prepare_gemm_a16wfp4_weight
```
Comment (fxmarty-amd) [inline vllm/model_executor/kernels/linear/mxfp4/aiter_a16wfp4.py]: Which AITER version adds this, which PR is it from? This is not in main yet, so you will need to have merged + AITER released + AITER version bumped in vLLM.

I would strongly recommend you to open a separate PR to add this backend

**Comment by LiuYinfeng01** [comment]
Thanks for the review. I split the two features as suggested:

This PR now only contains opt-in MXFP4 emulation AOT dequantization (default off). The AITER backend will follow after ROCm/aiter#4772 is released and adopted by vLLM.

**Review by fxmarty-amd — APPROVED** [review]
LGTM.

Please clarify the use case (example model + device) in the PR description & motivation in line with mxfp8 emulation

**Comment by fxmarty-amd** [comment]
@LiuYinfeng01 I already reviewed

**Review by shen-shanshan — APPROVED** [review]
Reviewed by @fxmarty-amd
